### DISCUSSION TITLE
How to run functional tests locally?

### DISCUSSION THREAD

**Discussion by lmilbaum** [python-gitlab/python-gitlab#2312] [Q&A]
OS: macOS 12.6

I've pocked into the code to figure this out. The Makefile in #2311 is a draft of what I figured out. It doesn't work for me yet.
Would appreaciate some insights.

**Comment by nejch** [comment]
Hey @lmilbaum thanks for looking into things. Generally, you only need `tox` and `docker` installed, as documented here:
https://github.com/python-gitlab/python-gitlab/blob/main/CONTRIBUTING.rst#running-unit-tests
https://github.com/python-gitlab/python-gitlab/blob/main/CONTRIBUTING.rst#running-integration-tests

What are the errors you are getting?

Note that the functional tests take quite a while to run, hence the `--keep-containers` flag, which should help with reusing a running gitlab container. It's quite a chunky container that takes around 2 minutes to spin up.

**Reply by nejch** [reply]
Also keep in mind that macOS might be limiting resources available to Docker, and gitlab is quite power hungry:
https://github.com/python-gitlab/python-gitlab/discussions/1519

Maybe you are hitting this issue and we need to document that, but would be good to know what you're seeing in the logs.

**Reply by lmilbaum** [reply]
Every function is failing on the same error `E       Exception: Timeout reached while waiting on service!`.
It seems that the container was not spinned up at all. `docker ps -a` returns empty list.

**Reply by nejch** [reply]
Ok, interesting. Does it spin up if you start compose manually @lmilbaum ?

```shell-session
docker-compose -f tests/functional/fixtures/docker-compose.yml up
```

**Reply by lmilbaum** [reply]
It does

**Reply by lmilbaum** [reply]
I run the functional target again and checked from a different shell that the gitlab-test container is spinned up. It does.  Could that the tests are starting to run before the container is ready and healthy?

**Reply by nejch** [reply]
Ah, great. Then I think you're really hitting the resource limit that I mentioned in the linked discussion. Could you check if your docker VM is limiting containers?

**Reply by lmilbaum** [reply]
docker VM? Not sure what it is. Let me check the documentation.

**Reply by nejch** [reply]
I think this might help you check the settings:
https://docs.docker.com/desktop/settings/mac/#resources
https://stackoverflow.com/questions/72299572/what-exactly-does-the-docker-desktop-memory-limit-apply-to

**Reply by lmilbaum** [reply]
<img width="941" alt="Screen Shot 2022-10-15 at 14 12 44" src="https://user-images.githubusercontent.com/7271026/195983565-4a0bbc17-ee65-43bf-8948-e01eecb72c88.png">

**Reply by nejch** [reply]
This should really be enough resources though, hmm.

Could you paste the full output of this command?

```shell
tox -e api_func_v4 -- --override-ini='log_cli=True'
```

**Reply by lmilbaum** [reply]
Sure. Will paste it soon. In the meanwhile, can you please clarify  when the environment spining is triggered.

**Reply by lmilbaum** [reply]
[my.log](https://github.com/python-gitlab/python-gitlab/files/9792161/my.log)

**Reply by nejch** [reply]
@lmilbaum the docker setup is handled by the pytest-docker plugin (https://github.com/avast/pytest-docker) so it is implemented as pytest fixtures.

We basically just add a `docker_compose_file` fixture and the plugin then calls `docker compose up` with that config file:
https://github.com/python-gitlab/python-gitlab/blob/9a6d197f9d2a88bdba8dab1f9abaa4e081a14792/tests/functional/conftest.py#L158-L160

The plugin then provides a few more fixtures for us to easily connect to the service via its container IP etc.

It looks to me like pytest can't connect to your GitLab container, so could also be a networking issue there. If you try http://127.0.0.1:8080, are you able to reach this as the container is spinning up or when running `docker-compose up` manually as shown above? It will initially show 502 and then you should see the login screen after a few minutes.

**Reply by lmilbaum** [reply]
Thanks for the spinning up clarification. Got it.
The root cause of my issue is the timeout. It started to work as soon as the timeout value was increased.

**Reply by nejch** [reply]
Ah great, glad you got it figured out! I wonder if increasing the docker resources even more might help. That does look like the container was struggling a bit. I'll add some docs on that as it's the second time this has happened on mac.

**Reply by lmilbaum** [reply]
Documenting this issue is a good idea. 
We could also consider waiting with running the functional tests until the gitlab service is healthy. And, if it doesn't become healthy in a specific time frame then issuing an error and exit.
That could help the developer in the mitigation process. WDYT?

**Reply by nejch** [reply]
I think we wanted to have some sensible timeout so that if the startup does actually fail for whatever reason (maybe even just `port already in use`), then the developer is not left waiting in the dark for too long. 300 seconds is already quite long IMO :)

I think it's just very unclear right now what happens at the moment. Instead, I think we could call `pytest.exit()` to prevent starting all the tests. We can probably do this here:

https://github.com/python-gitlab/python-gitlab/blob/8dbaa5cddef6d7527ded686553121173e33d2973/tests/functional/conftest.py#L262-L270

I've opened https://github.com/avast/pytest-docker/pull/88 so we can later narrow down the exception there.

**Comment by JohnVillalovos** [comment]
One thing that could be done is enable the verbose mode by default. So that it outputs the status of what is happening like is currently being done in the CI.

For me personally I enable the verbose mode when running locally.

**Reply by lmilbaum** [reply]
Could you please provide the steps for enabling the verbose mode?

**Reply by JohnVillalovos** [reply]
Sure. It is done here in the CI: https://github.com/python-gitlab/python-gitlab/blob/8dbaa5cddef6d7527ded686553121173e33d2973/.github/workflows/test.yml#L76

So I use this locally, to run the CLI functional tests. Would use similar arguments for the API functional tests:
`$ tox -e cli_func_v4 -- --override-ini='log_cli=True'`

**Reply by lmilbaum** [reply]
Thanks a lot
